lib/Sema/TypeCheckType.cpp

```diff
@@ -1180,6 +1180,32 @@ resolveTopLevelIdentTypeComponent(TypeChecker &TC, DeclContext *DC,
     return ErrorType::get(TC.Context);
   }
 
+  // Emit a warning about directly spelling
+  // ImplicitlyUnwrappedOptional rather than using a trailing '!'.
+  auto *IUODecl = TC.Context.getImplicitlyUnwrappedOptionalDecl();
+  if (currentDecl == IUODecl && TC.Context.isSwiftVersionAtLeast(4)) {
+    if (current->getNominalOrBoundGenericNominal() &&
+        options.contains(TR_AllowIUO)) {
+      auto *genericTyR = cast<GenericIdentTypeRepr>(comp);
```

Comment:
I'm not sure we've hit the point where `current->getNominalOrBoundGenericNominal()` implies that `cast<GenericIdentTypeRepr>(comp)` will succeed. I recommend gating this on `isa< GenericIdentTypeRepr>(comp)` instead.

Reply:
Ah, okay.